feat: TypeBox schemas, path metadata registry, and signalk-schema removal (#2325)
SignalK/signalk-server: docs/develop/websocket-protocol.md

## Endpoints

| Endpoint               | Purpose                  |
| ---------------------- | ------------------------ |
| `/signalk/v1/stream`   | Real-time data streaming |
| `/signalk/v1/playback` | Historical data playback |

Both endpoints use the standard WebSocket protocol (`ws://` or `wss://`).

> For the REST API (HTTP) documentation, see [REST APIs](./rest-api/README.md).
>
> For machine-readable channel documentation, see the [AsyncAPI viewer](/skServer/asyncapi/docs) (Documentation &rarr; AsyncAPI in the Admin UI sidebar).

---

## Connecting

### Basic Connection

Connect to the stream endpoint using any WebSocket client:

```
ws://localhost:3000/signalk/v1/stream
```

### Query Parameters

| Parameter          | Values                | Default  | Description                                   |
| ------------------ | --------------------- | -------- | --------------------------------------------- |
| `subscribe`        | `none`, `self`, `all` | `self`   | Initial subscription scope                    |
| `sendMeta`         | `all`                 | _(none)_ | Include metadata deltas for subscribed values |

> | `sendCachedValues` | `true`, `false`       | `true`   | Send cached values on connection              |

Reviewer: ```suggestion
| `sendCachedValues` | `true`, `false`       | `true`   | Send latest cached values on connection              |
```

| `token`            | JWT string            | _(none)_ | Authentication token                          |

_Example: Connect with no initial subscriptions (then subscribe selectively):_

```
ws://localhost:3000/signalk/v1/stream?subscribe=none
```

_Example: Subscribe to all vessels, include metadata:_

```
ws://localhost:3000/signalk/v1/stream?subscribe=all&sendMeta=all
```

### Playback Connection

The playback endpoint replays historical data:

```
ws://localhost:3000/signalk/v1/playback?startTime=2024-06-15T08:00:00Z&playbackRate=5
```

| Parameter      | Required | Description                               |
| -------------- | -------- | ----------------------------------------- |
| `startTime`    | Yes      | ISO 8601 timestamp to start playback from |
| `playbackRate` | No       | Speed multiplier (default: 1)             |

---

## Authentication

The server checks for authentication credentials in the following order:

1. **Query parameter:** `?token=<jwt>`
2. **HTTP header:** `Authorization: Bearer <jwt>` or `Authorization: JWT <jwt>`
3. **Cookie:** `JAUTHENTICATION=<jwt>`

> If no valid token is provided and the server has `allow_readonly` enabled, the connection is granted read-only access.

Reviewer: Since these are the same for http and ws connections I think we should have just a single place for authentication related documentation, except for the ws specific authentication method.

| ----------- | ------------------------------------- | ------------------------------------------------------------------------------- |
| NMEA2000    | `type`, `label`, `src`, `pgn`         | `{ "type": "NMEA2000", "label": "N2K", "src": "115", "pgn": 128267 }`           |
| NMEA0183    | `type`, `label`, `talker`, `sentence` | `{ "type": "NMEA0183", "label": "serial1", "talker": "GP", "sentence": "RMC" }` |
| SignalK     | `type`, `label`                       | `{ "type": "SignalK", "label": "courseApi" }`                                   |
| Plugin      | `label`                               | `{ "label": "my-plugin-id" }`                                                   |

### Value Types

Values for well-known paths are typically:

- **Numbers** — always in SI units (meters, radians, Kelvin, m/s, etc.)
- **Strings** — identifiers, names, states
- **Booleans** — binary states
- **Objects** — structured data (positions, notifications, etc.)
- **`null`** — path value cleared / resource deleted

_See the [Path Reference](/documentation/paths) for units and descriptions of all well-known paths._

> ### Vessel Name Delta

Reviewer: ```suggestion
### Context level properties
```

| `subscribe[].period`    | No       | 1000    | Update interval in milliseconds                                         |
| `subscribe[].format`    | No       | `delta` | Response format                                                         |
| `subscribe[].policy`    | No       | `ideal` | `instant` (every change), `ideal` (throttled), `fixed` (exact interval) |
| `subscribe[].minPeriod` | No       | 0       | Minimum interval between updates in milliseconds                        |

### Unsubscribe

```json
{
  "context": "vessels.self",
  "unsubscribe": [{ "path": "navigation.speedThroughWater" }]
}
```

Unsubscribe from everything:

```json
{
  "context": "*",
  "unsubscribe": [{ "path": "*" }]
}
```

### Path Discovery with `announceNewPaths`

To discover available paths without subscribing to continuous updates for all of them:

```json
{
  "context": "*",
  "announceNewPaths": true,
  "subscribe": [
    {
      "path": "navigation.position",
      "period": 1000
    }
  ]
}
```

When `announceNewPaths` is `true`:

1. The server sends cached values for **all** existing paths matching the context (once each)

> 2. When a new path appears later (e.g. new sensor comes online), the server announces it once

Reviewer: ```suggestion
2. When a new path appears later (e.g. new sensor comes online), the server sends its value value once, announcing its existence.
```
